Sources/SourceKitLSP/SourceKitServer.swift

```diff
@@ -527,7 +548,24 @@ extension SourceKitServer {
       executeCommandProvider: ExecuteCommandOptions(
         commands: builtinSwiftCommands // FIXME: Clangd commands?
       )
-    )))
+    )
+  }
+
+  func registerCapabilities(
+    for server: ServerCapabilities,
+    languages: [Language],
+    registry: CapabilityRegistry
+  ) {
+    guard let completionOptions = server.completionProvider else { return }
```

Comment:
The way this method is structured makes it a bit tricky to extend when we eventually want to add other kinds of dynamic capabilities in addition to code-completion.  I wonder if we should move most of this code down to the registry itself (which ultimately is the place that knows what is supported) and just have this method be responsible for sending the message back to the client.

Additionally, have you thought about the tradeoffs between the design you currently have where the registry is simple and single-threaded, but you need to remember to always change the registrations via this API on the server (or else they will not be communicated back to the client), vs. having the registry keep a callback function that it can use to send updates to the client directly?

Reply:
Yeah that seems like it would be good a idea, either using a protocol or callback. Depending on the spec we might need to unregister capabilities before registering new ones, I opened this [issue](https://github.com/microsoft/language-server-protocol/issues/1227) about that but it hasn't gotten a response, will likely need to test it out with VS Code to see how VS Code behaves.

Reply:
Looks good, thanks!